web/src/user_groups.ts

```diff
@@ -311,6 +311,31 @@ export function get_recursive_group_members(target_user_group: UserGroup): Set<n
     return members;
 }
 
+export function get_potential_subgroups(target_user_group_id: number): UserGroup[] {
+    const target_user_group = get_user_group_from_id(target_user_group_id);
+    const already_subgroup_ids = target_user_group.direct_subgroup_ids;
+    return get_all_realm_user_groups().filter((user_group) => {
+        if (user_group.id === target_user_group.id) {
+            return false;
+        }
+
+        if (already_subgroup_ids.has(user_group.id)) {
+            return false;
+        }
+
+        if (user_group.direct_subgroup_ids.has(target_user_group.id)) {
+            return false;
+        }
+
+        const recursive_subgroup_ids = get_recursive_subgroups(user_group);
```

Comment:
I think `user_group.direct_subgroup_ids` is already in this list, so we don't need the previous `return false` block?

```
export function get_recursive_subgroups(target_user_group: UserGroup): Set<number> | undefined {        
    // Correctness of this algorithm relying on the ES6 Set                                             
    // implementation having the property that a `for of` loop will                                     
    // visit all items that are added to the set during the loop.                                       
    const subgroup_ids = new Set(target_user_group.direct_subgroup_ids);                                
    for (const subgroup_id of subgroup_ids) {                                                           
        const subgroup = user_group_by_id_dict.get(subgroup_id);                                        
```

Reply:
Deleted the unnecessary `if` statement and confirmed the tests still pass. Also added a comment on this function noting that it could be optimized if we cached supergroup references, but given the numbers of groups with subgroups that are realistic, it does not seem like a priority at all to do that refactor.

Reply:
I thought that keeping the `direct_subgroup_ids` could help in avoiding calculating the recursive subgroups for some cases.